pjmedia/src/pjmedia-codec/lyra.cpp

```diff
@@ -0,0 +1,715 @@
+    }
+
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_alloc_codec(pjmedia_codec_factory *factory,
+                                    const pjmedia_codec_info *id,
+                                    pjmedia_codec **p_codec)
+{
+    pj_pool_t *pool;
+    struct lyra_data *lyra_data;
+    pjmedia_codec *codec;
+    pj_status_t status;
+
+    PJ_ASSERT_RETURN(factory && id && p_codec, PJ_EINVAL);
+    PJ_ASSERT_RETURN(factory == &lyra_factory.base, PJ_EINVAL);
+
+    pool = pjmedia_endpt_create_pool(lyra_factory.endpt, "lyra%p", 2000, 2000);
+    PJ_ASSERT_RETURN(pool != NULL, PJ_ENOMEM);
+
+    lyra_data = PJ_POOL_ZALLOC_T(pool, struct lyra_data);
+    codec = PJ_POOL_ZALLOC_T(pool, pjmedia_codec);
+
+    status = pj_mutex_create_simple(pool, "lyra_mutex", &lyra_data->mutex);
+    if (status != PJ_SUCCESS) {
+        pj_pool_release(pool);
+        return status;
+    }
+
+    lyra_data->pool = pool;
+    lyra_data->dec_bit_rate = lyra_cfg.bit_rate;
+
+    codec->op = &lyra_op;
+    codec->codec_data = lyra_data;
+    codec->factory = factory;
+
+    *p_codec = codec;
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_dealloc_codec(pjmedia_codec_factory *factory,
+                                      pjmedia_codec *codec)
+{
+    struct lyra_data *lyra_data;
+
+    PJ_ASSERT_RETURN(factory && codec, PJ_EINVAL);
+    PJ_UNUSED_ARG(factory);
+    PJ_ASSERT_RETURN(factory == &lyra_factory.base, PJ_EINVAL);
+
+    lyra_data = (struct lyra_data*)codec->codec_data;
+    if (lyra_data) {
+        pj_mutex_destroy(lyra_data->mutex);
+        lyra_data->mutex = NULL;
+        pj_pool_release(lyra_data->pool);
+    }
+
+    return PJ_SUCCESS;
+}
+
+/* Prototypes for lyra implementation. */
+static pj_status_t lyra_codec_init(pjmedia_codec *codec,
+                                   pj_pool_t *pool)
+{
+    PJ_UNUSED_ARG(codec);
+    PJ_UNUSED_ARG(pool);
+    return PJ_SUCCESS;
+}
+
+static pj_status_t lyra_codec_open(pjmedia_codec *codec,
+                                   pjmedia_codec_param *attr)
+{
+    ghc::filesystem::path model_path = lyra_cfg.model_path.ptr;
+    struct lyra_data *lyra_data = (struct lyra_data*)codec->codec_data;
+
+    pj_mutex_lock(lyra_data->mutex);
+
+    PJ_LOG(4, (THIS_FILE, "Codec opening, model_path=%.*s",
+            (int)lyra_cfg.model_path.slen, lyra_cfg.model_path.ptr));
+
+    /* Get encoder bit_rate */
+    for (unsigned i = 0; i < attr->setting.enc_fmtp.cnt; ++i) {
+        if (pj_strcmp2(&attr->setting.enc_fmtp.param[i].name, 
+                       BIT_RATE_STR) == 0)
+        {
+            lyra_data->enc_bit_rate = (pj_uint16_t)
+                              pj_strtoul(&attr->setting.enc_fmtp.param[i].val);
+            break;
+        }
+    }
+    if (lyra_data->enc_bit_rate == 0 ||
+        (lyra_data->enc_bit_rate != 3200 && lyra_data->enc_bit_rate != 6000 &&
+         lyra_data->enc_bit_rate != 9200))
+    {
+        lyra_data->enc_bit_rate = PJMEDIA_CODEC_LYRA_DEFAULT_BIT_RATE;
+    }

```

Comment:
Perhaps better to also get the decoder bitrate from SDP as the app is allowed to modify the SDP?